sound/soc/sof/intel/hda-codec.c

```diff
@@ -117,6 +117,66 @@ int hda_codec_probe_bus(struct snd_sof_dev *sdev)
 }
 EXPORT_SYMBOL(hda_codec_probe_bus);
 
+static void hda_pci_update_bits(struct pci_dev *pci, u32 offset,
+				u32 mask, u32 value)
+{
+	u32 new, ret;
+
+	pci_read_config_dword(pci, offset, &ret);
+	new = (ret & (~mask)) | (value & mask);
+	pci_write_config_dword(pci, offset, new);
+}
+
+int hda_codec_detect(struct pci_dev *pci, struct hdac_bus *bus, int init_chip)
+{
+	int ret;
+
+	if (!init_chip) {
+		snd_hdac_ext_bus_init(bus, &pci->dev, NULL, NULL, NULL);
+
+		bus->addr = pci_resource_start(pci, 0);
+		bus->remap_addr = pci_ioremap_bar(pci, 0);
+		if (!bus->remap_addr) {
+			dev_err(bus->dev, "error: ioremap error\n");
+			return 0;
+		}
+	}
+
+	/* i915 exposes a HDA codec for HDMI audio */
+	ret = snd_hdac_i915_init(bus);
+	if (ret < 0)
+		dev_dbg(&pci->dev, "warning: failed to init HDMI");
+
+	snd_hdac_display_power(bus, HDA_MAX_CODECS, true);
+
+	/* disable audio dsp clock gating and power gating */
+	hda_pci_update_bits(pci, PCI_CGCTL, PCI_CGCTL_MISCBDCGE_MASK, 0);
+
+	if (init_chip)
+		ret = snd_hdac_bus_init_chip(bus, true);
+	else
+		/* reset codec link to detect hdac codecs */
+		ret = snd_hdac_bus_reset_link(bus, true);
+
+	if (ret < 0)
+		return ret;
```

Comment:
Bam. your error flow did not turn i915 power off. this needs to be reworked.

Reply:
yes, I will refine it